teps/0121-refine-retries-for-taskruns-and-customruns.md

```diff
@@ -156,40 +353,174 @@ Three sub-options about the way to implement `retriesStatus`:
 - Move logic for `retries` to PipelineRun reconciler and create new `TaskRun`s and `Runs` at each attempt.
 - Remove `retriesStatus` from TaskRun & CustomRun
 
-**Benefits:**
+### Details
 
-- Consistent interface for `retries`
-- Custom task controller developers get a default implementation of retries for free (by embedding in a pipeline)
-- "Pipelines in pipeline" can be retried the same as the other resources
-- Improve the retries of TaskRuns created from PipelineTasks by using separate TaskRuns for each retry
-- No changes to the PipelineRun API (not in the spec at least)
-- No changes to the TaskRun API (not in the spec at least)
+In this option, if a Pipeline Task specifies `retries`, the PipelineRun controller will create a new instance of the CustomRun or TaskRun if the previous one has failed.
+It's already possible to have multiple TaskRuns or CustomRuns for a Pipeline Task, if it's matrixed.
+The same strategy can be used for reporting retried PipelineTasks, for example:
 
-**Concerns:**
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+spec:
+  ...
+status:
+  childReferences:
+    - apiVersion: tekton.dev/v1beta1
+      kind: TaskRun
+      name: foo-attempt-0
+      pipelineTaskName: foo
+    - apiVersion: tekton.dev/v1beta1
+      kind: TaskRun
+      name: foo-attempt-1
+      pipelineTaskName: foo
+```
 
-- API Change for `Run` and `CustomRun` (need to remove `retries` & `retriesStatus`)
-  - We are moving Custom Task Run from alpha (Run) to beta (CustomRun) (see [TEP-0114](https://github.com/tektoncd/community/blob/main/teps/0114-custom-tasks-beta.md)), which is a great timing for us to remove fields from `Run`.
-- Dashboard and CLI may need extra works if we remove `retriesStatus`
-- Standalone `TaskRun` can't retry on its own.
-- It's not quite user-friendly if a CustomRun controller implements its own retry strategy, for example:
+TBD:
+- What would this look like for PipelineTasks that are both matrixed and retryable?
```

Comment:
Heh, this would require some extra coding in the parent controller, to take care of only retrying failed attempts from the matrix. Retries implemented by the child resource would work with matrix out of the box.